README.md
# merge_images in flutter

<a href="https://https://pub.dev/packages/merge_images"><img src="https://img.shields.io/pub/v/merge_images.svg"/></a>  </a><a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-yellow.svg"/></a>

[中文文档](README(CH).md)

## Intro 
Merge images (Image stitching) in vertical or horizontal direction.
![Preview](example/preview/preview1.png)

## Features
* Support vertical and horizontal direction.
* Provide a helper to merge images in code and get a result image, and a widget to automatically merge and show images.
* Automatically scale the image to fit other images (fit width in vertical, and height in horizontal).

## Usage
#### ImagesMergeHelper
Use this helper to merge images in code.

``` dart
ui.Image image = await ImagesMergeHelper.margeImages(
[assetImage1,assetImage2,providerImage],///required,images list
   fit: true,///scale image to fit others
   direction: Axis.vertical,///direction of images
   backgroundColor: Colors.black26);///background color
```
Besides, it provider some functions to do image format conversion:
``` dart
///ui.Image to Uint8List
Uint8List bytes = await ImagesMergeHelper.imageToUint8List(image);
///ui.Image to File
File file = await ImagesMergeHelper.imageToFile(image);
///Uint8List to ui.Image
ui.Image image = await ImagesMergeHelper.uint8ListToImage(imageBytes);
///file to ui.Image
ui.Image image = await ImagesMergeHelper.loadImageFromFile(file);
///asset to ui.Image
ui.Image image = await ImagesMergeHelper.loadImageFromAsset(assetPath);
///ImageProvider to ui.Image
ui.Image image = await ImagesMergeHelper.loadImageFromProvider(imageProvider);

```
#### ImageMerge
Use this widget to automatically merge and show images.
``` dart
ImagesMerge(
  imageList,///required,images list
  direction: Axis.vertical,///direction
  backgroundColor: Colors.black26,///background color
  fit: false,///scale image to fit others
  controller: captureController,///controller to get screen shot
),
```
Besides, the widget was wrapped with RepaintBoundary, you can simply passing an instance of CaptureController and get the screen shot by calling capture(). 
``` dart
///get capture of widget by RepaintBoundary
 getCapture() async{
  Uint8List bytes = await captureController.capture();
  }
```
## LICENSE
MIT License



CHANGELOG.md
## [0.0.1]

* initial release

## [1.0.0]

* resolve dependency, format code

## [1.0.1]

* remove unused files

## [1.0.2]

* update ImagesMergeHelper

## [1.0.3]

* fix bugs of ImagesMergeHelper

## [2.0.0-nullsafety]
* migrate to null-safety

example/lib/main.dart
import 'package:flutter/material.dart';

import 'helper_page.dart';
import 'widget_page.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Example Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElevatedButton(
                onPressed: _toWidget,
                child: Text(
                  "ImagesMerge Widget",
                  style: Theme.of(context).textTheme.headline6,
                )),
            SizedBox(
              height: 30,
            ),
            ElevatedButton(
                onPressed: _toHelper,
                child: Text(
                  "ImagesMerge Helper",
                  style: Theme.of(context).textTheme.headline6,
                )),
          ],
        ),
      ),
    );
  }

  _toWidget() {
    Navigator.push(context,
        MaterialPageRoute(builder: (context) => ImagesMergeWidgetPage()));
  }

  _toHelper() {
    Navigator.push(context,
        MaterialPageRoute(builder: (context) => ImagesMergeHelperPage()));
  }
}


lib/src/images_merge.dart
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'dart:ui' as ui;
import 'package:flutter/rendering.dart';

part "merge_painter.dart";

///
///@author xiaozhizhong
///@date 2020/4/1
///@description images merge widget
///
// ignore: must_be_immutable
class ImagesMerge extends StatelessWidget {
  ImagesMerge(this.imageList,
      {this.direction = Axis.vertical,
      this.controller,
      this.fit = true,
      this.backgroundColor});

  ///List of images list, content must be ui.Image.
  ///If you have another format of image, you can transfer it to ui.Image
  ///by [ImagesMergeHelper].
  final List<ui.Image> imageList;

  ///Merge direction, default to vertical.
  final Axis direction;

  ///Whether to Scale the pictures to same width/height when pictures has
  ///different width/height,
  ///Fit width when direction is vertical, and fit height when horizontal.
  ///Default to true.
  final bool fit;

  ///background color
  final Color? backgroundColor;

  ///Controller to capture screen.
  final CaptureController? controller;

  int totalWidth = 0;
  int totalHeight = 0;

  double scale = 1;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraint) {
        _calculate(constraint);
        return RepaintBoundary(
          key: controller?.key ?? ValueKey(0),
          child: ClipRRect(
            child: Container(
              color: backgroundColor,
              child: CustomPaint(
                painter: _MergePainter(imageList, direction, fit, scale),
                size: Size(totalWidth.toDouble(), totalHeight.toDouble()),
              ),
            ),
          ),
        );
      },
    );
  }

  ///calculating width and height of canvas
  _calculate(BoxConstraints constraint) {
    //calculate the max width/height of images
    imageList.forEach((image) {
      if (direction == Axis.vertical) {
        if (totalWidth < image.width) totalWidth = image.width;
      } else {
        if (totalHeight < image.height) totalHeight = image.height;
      }
    });
    //calculate the constraint of parent
    if (direction == Axis.vertical &&
        constraint.hasBoundedWidth &&
        totalWidth > constraint.maxWidth) {
      scale = constraint.maxWidth / totalWidth;
      totalWidth = constraint.maxWidth.floor();
    } else if (direction == Axis.horizontal &&
        constraint.hasBoundedHeight &&
        totalHeight > constraint.maxHeight) {
      scale = constraint.maxHeight / totalHeight;
      totalHeight = constraint.maxHeight.floor();
    }
    //calculate the opposite
    imageList.forEach((image) {
      if (direction == Axis.vertical) {
        if (image.width < totalWidth && !fit) {
          totalHeight += image.height;
        } else {
          if (!fit)
            totalHeight += (image.height * scale).floor();
          else
            totalHeight += (image.height * totalWidth / image.width).floor();
        }
      } else {
        if (image.height < totalHeight && !fit) {
          totalWidth += image.width;
        } else {
          if (!fit)
            totalWidth += (image.width * scale).floor();
          else
            totalWidth += (image.width * totalHeight / image.height).floor();
        }
      }
    });
  }
}

/// Screen shot capture controller
class CaptureController {
  CaptureController();

  final GlobalKey key = GlobalKey();

  ///capture the screen shot by RepaintBoundary
  Future<Uint8List?> capture() async {
    try {
      RenderRepaintBoundary boundary =
          key.currentContext!.findRenderObject() as RenderRepaintBoundary;
      double dpr = ui.window.devicePixelRatio;
      ui.Image image = await boundary.toImage(pixelRatio: dpr);
      ByteData? byteData = await image.toByteData(format: ui.ImageByteFormat.png);
      Uint8List? pngBytes = byteData?.buffer.asUint8List();
      return pngBytes;
    } catch (e) {
      print(e);
    }
    return null;
  }
}


lib/src/merge_painter.dart
part of "images_merge.dart";

///
///@author xiaozhizhong
///@date 2020/4/1
///@description painter
///
class _MergePainter extends CustomPainter {
  _MergePainter(this.imageList, this.direction, this.fit, this.scale);

  final List<ui.Image> imageList;
  final Axis direction;
  final bool fit;
  final double scale;

  @override
  void paint(Canvas canvas, Size size) {
    double dx = 0;
    double dy = 0;
    double totalWidth = size.width;
    double totalHeight = size.height;
    Paint paint = Paint();
    imageList.forEach((image) {
      //scale the image to same width/height
      late double imageHeight;
      late double imageWidth;
      double dxScale = dx;
      double dyScale = dy;
      if (direction == Axis.vertical) {
        if (image.width < totalWidth && !fit) {
          canvas.drawImage(image, Offset(dx, dy), paint);
        } else {
          canvas.save();
          if (!fit) {
            imageHeight = image.height * scale;
            canvas.scale(imageHeight / image.height);
          } else {
            canvas.scale(totalWidth / image.width);
            imageHeight = image.height * totalWidth / image.width;
          }

          dyScale *= image.height / imageHeight;
          canvas.drawImage(image, Offset(dxScale, dyScale), paint);
          canvas.restore();
        }
      } else {
        if (image.height < totalHeight && !fit) {
          canvas.drawImage(image, Offset(dx, dy), paint);
        } else {
          canvas.save();
          if (!fit) {
            imageWidth = image.width * scale;
            canvas.scale(imageWidth / image.width);
          } else {
            canvas.scale(totalHeight / image.height);
            imageWidth = image.width * totalHeight / image.height;
          }
          dxScale *= image.width / imageWidth;
          canvas.drawImage(image, Offset(dxScale, dyScale), paint);
          canvas.restore();
        }
      }
      if (direction == Axis.vertical) {
        dy += imageHeight;
      } else {
        dx += imageWidth;
      }
    });
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return oldDelegate != this;
  }
}


example/lib/widget_page.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:merge_images/merge_images.dart';
import 'dart:ui' as ui;

import 'image_preview_page.dart';

///
///@author xiaozhizhong
///@date 2020/4/2
///@description images merge widget example page
///
class ImagesMergeWidgetPage extends StatefulWidget {
  @override
  _ImagesMergeWidgetPageState createState() => _ImagesMergeWidgetPageState();
}

class _ImagesMergeWidgetPageState extends State<ImagesMergeWidgetPage> {
  var imageList = <ui.Image>[];
  late ui.Image assetImage1;
  late ui.Image assetImage2;
  late ui.Image providerImage;

  final networkImagePath = "http://img.article.pchome.net/00/29/20/31/pic_lib/s960x639/Sakura_28s960x639.jpg";

  final captureController = CaptureController();

  @override
  void initState() {
    super.initState();
    loadImage();
  }

  loadImage() async {
    assetImage1 = await ImagesMergeHelper.loadImageFromAsset("assets/sunset.jpeg");
    assetImage2 = await ImagesMergeHelper.loadImageFromAsset("assets/bridge.jpg");
    providerImage = await ImagesMergeHelper.loadImageFromProvider(NetworkImage(networkImagePath));
    setState(() {
      imageList.add(assetImage1);
      imageList.add(assetImage2);
      imageList.add(providerImage);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Widget Page"),
      ),
      body: Container(
        width: double.infinity,
        height: double.infinity,
        padding: EdgeInsets.all(30),
        child: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              Image.asset(
                "assets/sunset.jpeg",
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              SizedBox(
                height: 10,
              ),
              Image.asset(
                "assets/bridge.jpg",
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              SizedBox(
                height: 10,
              ),
              Image.network(
                networkImagePath,
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Text("merge(vertical&fit)"),
              ),
              Offstage(
                offstage: imageList.isEmpty,
                child: Container(
                  width: 100,
                  child: ImagesMerge(
                    imageList,
                    direction: Axis.vertical,
                    fit: true,
                  ),
                ),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Text("merge(vertical)"),
              ),
              Offstage(
                offstage: imageList.isEmpty,
                child: Container(
                  width: 100,
                  child: ImagesMerge(
                    imageList,
                    direction: Axis.vertical,
                    backgroundColor: Colors.black26,
                    fit: false,
                  ),
                ),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Text("merge(horizontal&fit)"),
              ),
              Offstage(
                offstage: imageList.isEmpty,
                child: Container(
                  height: 70,
                  child: ImagesMerge(
                    imageList,
                    direction: Axis.horizontal,
                    fit: true,
                  ),
                ),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Text("merge(horizontal)"),
              ),
              Offstage(
                offstage: imageList.isEmpty,
                child: Container(
                  height: 70,
                  child: ImagesMerge(
                    imageList,
                    direction: Axis.horizontal,
                    fit: false,
                    backgroundColor: Colors.black26,
                    controller: captureController,
                  ),
                ),
              ),
              ElevatedButton(
                  onPressed: () => getCapture(context), child: Text("Capture", style: Theme.of(context).textTheme.headline6))
            ],
          ),
        ),
      ),
    );
  }

  ///get capture of widget by RepaintBoundary
  getCapture(context) async {
    Uint8List? bytes = await captureController.capture();
    if (bytes == null) return;
    Navigator.push(context, MaterialPageRoute(builder: (context) => Preview(bytes)));
  }
}


lib/src/images_merge_helper.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'dart:ui' as ui;

import 'package:flutter/services.dart';
import 'package:path_provider/path_provider.dart';

///
///@author xiaozhizhong
///@date 2020/4/1
///@description images merge helper
///

class ImagesMergeHelper {
  ///Merge images from a list of ui.Image
  ///[imageList] list of ui.Image
  ///[direction] merge direction, default to vertical
  ///[fit] Whether to Scale the pictures to same width/height when pictures has
  ///different width/height,
  /// Fit width when direction is vertical, and fit height when horizontal.
  /// Default to true.
  ///[backgroundColor] background color of picture
  static Future<ui.Image> margeImages(List<ui.Image> imageList,
      {Axis direction = Axis.vertical, bool fit = true, Color? backgroundColor}) {
    int maxWidth = 0;
    int maxHeight = 0;
    //calculate max width/height of image
    imageList.forEach((image) {
      if (direction == Axis.vertical) {
        if (maxWidth < image.width) maxWidth = image.width;
      } else {
        if (maxHeight < image.height) maxHeight = image.height;
      }
    });
    int totalHeight = maxHeight;
    int totalWidth = maxWidth;
    ui.PictureRecorder recorder = ui.PictureRecorder();
    final paint = Paint();
    Canvas canvas = Canvas(recorder);
    double dx = 0;
    double dy = 0;
    //set background color
    if (backgroundColor != null) canvas.drawColor(backgroundColor, BlendMode.srcOver);
    //draw images into canvas
    imageList.forEach((image) {
      double scaleDx = dx;
      double scaleDy = dy;
      double imageHeight = image.height.toDouble();
      double imageWidth = image.width.toDouble();
      if (fit) {
        //scale the image to same width/height
        canvas.save();
        if (direction == Axis.vertical && image.width != maxWidth) {
          canvas.scale(maxWidth / image.width);
          scaleDy *= imageWidth / maxWidth;
          imageHeight *= maxWidth / imageWidth;
        } else if (direction == Axis.horizontal && image.height != maxHeight) {
          canvas.scale(maxHeight / image.height);
          scaleDx *= imageHeight / maxHeight;
          imageWidth *= maxHeight / imageHeight;
        }
        canvas.drawImage(image, Offset(scaleDx, scaleDy), paint);
        canvas.restore();
      } else {
        //draw directly
        canvas.drawImage(image, Offset(dx, dy), paint);
      }
      //accumulate dx/dy
      if (direction == Axis.vertical) {
        dy += imageHeight;
        totalHeight += imageHeight.floor();
      } else {
        dx += imageWidth;
        totalWidth += imageWidth.floor();
      }
    });
    //output image
    return recorder.endRecording().toImage(totalWidth, totalHeight);
  }

  ///transfer ui.Image to Unit8List
  ///[image]
  ///[format] default to png
  static Future<Uint8List?> imageToUint8List(ui.Image image, {ui.ImageByteFormat format = ui.ImageByteFormat.png}) async {
    ByteData? byteData = await image.toByteData(format: format);
    return byteData?.buffer.asUint8List();
  }

  ///transfer ui.Image to File
  ///[image]
  ///[path] path to store temporary file, will use [ getTemporaryDirectory ]
  ///by path_provider if null
  static Future<File?> imageToFile(ui.Image image, {String? path}) async {
    Uint8List? byte = await imageToUint8List(image);
    if(byte==null) return null;
    final directory = path ?? (await getTemporaryDirectory()).path;
    String fileName = DateTime.now().toIso8601String();
    String fullPath = '$directory/$fileName.png';
    File imgFile = new File(fullPath);
    return await imgFile.writeAsBytes(byte);
  }

  ///transfer Unit8List to ui.Image
  ///[bytes] Uint8List bytes
  static Future<ui.Image> uint8ListToImage(Uint8List bytes) async {
    ImageProvider provider = MemoryImage(bytes);
    return await loadImageFromProvider(provider);
  }

  ///load ui.Image from File
  ///[file] Image file
  static Future<ui.Image> loadImageFromFile(File file) async {
    Uint8List bytes = file.readAsBytesSync();
    return await uint8ListToImage(bytes);
  }

  ///load ui.Image from asset
  ///[asset] asset path
  static Future<ui.Image> loadImageFromAsset(String asset) async {
    ByteData data = await rootBundle.load(asset);
    var codec = await ui.instantiateImageCodec(data.buffer.asUint8List());
    ui.FrameInfo fi = await codec.getNextFrame();
    return fi.image;
  }

  ///load ui.Image from ImageProvider
  ///[provider] ImageProvider
  static Future<ui.Image> loadImageFromProvider(
    ImageProvider provider, {
    ImageConfiguration config = ImageConfiguration.empty,
  }) async {
    Completer<ui.Image> completer = Completer<ui.Image>();
    late ImageStreamListener listener;
    ImageStream stream = provider.resolve(config);
    listener = ImageStreamListener((ImageInfo frame, bool sync) {
      final ui.Image image = frame.image;
      completer.complete(image);
      stream.removeListener(listener);
    });
    stream.addListener(listener);
    return completer.future;
  }
}


example/lib/helper_page.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'dart:ui' as ui;

import 'package:merge_images/merge_images.dart';

import 'image_preview_page.dart';

///
///@author xiaozhizhong
///@date 2020/4/2
///@description images merge helper example page
///
class ImagesMergeHelperPage extends StatefulWidget {
  @override
  _ImagesMergeHelperPageState createState() => _ImagesMergeHelperPageState();
}

class _ImagesMergeHelperPageState extends State<ImagesMergeHelperPage> {
  ui.Image? assetImage1;
  ui.Image? assetImage2;
  ui.Image? providerImage;

  final networkImagePath = "http://img.article.pchome.net/00/29/20/31/pic_lib/s960x639/Sakura_28s960x639.jpg";

  @override
  void initState() {
    super.initState();
    loadImage();
  }

  loadImage() async {
    assetImage1 = await ImagesMergeHelper.loadImageFromAsset("assets/sunset.jpeg");
    assetImage2 = await ImagesMergeHelper.loadImageFromAsset("assets/bridge.jpg");
    providerImage = await ImagesMergeHelper.loadImageFromProvider(NetworkImage(networkImagePath));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Helper Page"),
      ),
      body: Container(
        width: double.infinity,
        height: double.infinity,
        padding: EdgeInsets.all(30),
        child: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              Image.asset(
                "assets/sunset.jpeg",
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              SizedBox(
                height: 10,
              ),
              Image.asset(
                "assets/bridge.jpg",
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              SizedBox(
                height: 10,
              ),
              Image.network(
                networkImagePath,
                scale: 10,
                fit: BoxFit.scaleDown,
              ),
              SizedBox(
                height: 30,
              ),
              ElevatedButton(
                  onPressed: () => _merge(fit: true, direction: Axis.vertical),
                  child: Text(
                    "vertical & fit",
                    style: Theme.of(context).textTheme.headline6,
                  )),
              ElevatedButton(
                  onPressed: () => _merge(fit: false, direction: Axis.vertical),
                  child: Text(
                    "vertical",
                    style: Theme.of(context).textTheme.headline6,
                  )),
              ElevatedButton(
                  onPressed: () => _merge(fit: true, direction: Axis.horizontal),
                  child: Text(
                    "horizontal & fit",
                    style: Theme.of(context).textTheme.headline6,
                  )),
              ElevatedButton(
                  onPressed: () => _merge(fit: false, direction: Axis.horizontal),
                  child: Text(
                    "horizontal",
                    style: Theme.of(context).textTheme.headline6,
                  )),
            ],
          ),
        ),
      ),
    );
  }

  ///merge images using ImagesMergeHelper and preview
  _merge({required bool fit, required Axis direction}) async {
    ui.Image image = await ImagesMergeHelper.margeImages([assetImage1!, assetImage2!, providerImage!],
        fit: fit, direction: direction, backgroundColor: Colors.black26);
    Uint8List? bytes = await ImagesMergeHelper.imageToUint8List(image);
    if (bytes == null) return;
    Navigator.push(context, MaterialPageRoute(builder: (context) => Preview(bytes)));
  }
}
